### PR TITLE
Brute-force correctness checker for our LLVM semantics versus `opt`'s peephole optimizer

### PR DESCRIPTION
**Pull request by bollu**
We add the following:

1. A python script that generates an LLVM file, runs `opt -instcombine` on it, extracts out the optimized values, and writes it out to e a CSV fule,
2. A lean script that runs our LLVM semantics and builds a matching CSV fule
3. A `run.sh` script that diffs the two CSV files to check for semantics equivalence
4. A CI runner that ensures that the semantics continue to agree.

### CODE REVIEW

**Comment by bollu** [comment]
We seem to have some difference in `srem` and `sdiv`: 
```diff
+ diff generated-llvm-optimized-data.csv generated-ssa-llvm-semantics.csv
591c591
< sdiv, 1, 0, 1, <none>, false
---
> sdiv, 1, 0, 1, <none>, poison
593c593
< sdiv, 1, 1, 1, <none>, true
---
> sdiv, 1, 1, 1, <none>, poison
599c599
< sdiv, 2, 1, 1, <none>, 1
---
> sdiv, 2, 1, 1, <none>, poison
603,604c603,604
< sdiv, 2, 2, 1, <none>, -2
< sdiv, 2, 2, 2, <none>, 1
---
> sdiv, 2, 2, 1, <none>, poison
> sdiv, 2, 2, 2, <none>, poison
608,609c608,609
< sdiv, 2, 3, 2, <none>, 0
< sdiv, 2, 3, 3, <none>, 1
---
> sdiv, 2, 3, 2, <none>, poison
> sdiv, 2, 3, 3, <none>, poison
635c635
< sdiv, 3, 3, 1, <none>, 3
---
> sdiv, 3, 3, 1, <none>, poison
643c643
< sdiv, 3, 4, 1, <none>, -4
---
> sdiv, 3, 4, 1, <none>, poison
645c645
< sdiv, 3, 4, 3, <none>, -1
---
> sdiv, 3, 4, 3, <none>, -2
647c647
< sdiv, 3, 4, 5, <none>, 1
---
> sdiv, 3, 4, 5, <none>, 2
652c652
< sdiv, 3, 5, 2, <none>, -1
---
> sdiv, 3, 5, 2, <none>, -2
654c654
< sdiv, 3, 5, 4, <none>, 0
---
> sdiv, 3, 5, 4, <none>, 1
656,657c656,657
< sdiv, 3, 5, 6, <none>, 1
< sdiv, 3, 5, 7, <none>, 3
---
> sdiv, 3, 5, 6, <none>, 2
> sdiv, 3, 5, 7, <none>, poison
661,663c661,663
< sdiv, 3, 6, 3, <none>, 0
< sdiv, 3, 6, 4, <none>, 0
< sdiv, 3, 6, 5, <none>, 0
---
> sdiv, 3, 6, 3, <none>, -1
> sdiv, 3, 6, 4, <none>, 1
> sdiv, 3, 6, 5, <none>, 1
668,672c668,672
< sdiv, 3, 7, 2, <none>, 0
< sdiv, 3, 7, 3, <none>, 0
< sdiv, 3, 7, 4, <none>, 0
< sdiv, 3, 7, 5, <none>, 0
< sdiv, 3, 7, 6, <none>, 0
---
> sdiv, 3, 7, 2, <none>, -1
> sdiv, 3, 7, 3, <none>, -1
> sdiv, 3, 7, 4, <none>, 1
> sdiv, 3, 7, 5, <none>, 1
> sdiv, 3, 7, 6, <none>, 1
759c759
< srem, 1, 0, 1, <none>, false
---
> srem, 1, 0, 1, <none>, poison
761c761
< srem, 1, 1, 1, <none>, false
---
> srem, 1, 1, 1, <none>, poison
767c767
< srem, 2, 1, 1, <none>, 0
---
> srem, 2, 1, 1, <none>, poison
771,772c771,772
< srem, 2, 2, 1, <none>, 0
< srem, 2, 2, 2, <none>, 0
---
> srem, 2, 2, 1, <none>, poison
> srem, 2, 2, 2, <none>, poison
775,777c775,777
< srem, 2, 3, 1, <none>, 0
< srem, 2, 3, 2, <none>, -1
< srem, 2, 3, 3, <none>, 0
---
> srem, 2, 3, 1, <none>, -1
> srem, 2, 3, 2, <none>, poison
> srem, 2, 3, 3, <none>, poison
803c803
< srem, 3, 3, 1, <none>, 0
---
> srem, 3, 3, 1, <none>, poison
811,812c811,812
< srem, 3, 4, 1, <none>, 0
< srem, 3, 4, 2, <none>, 0
---
> srem, 3, 4, 1, <none>, poison
> srem, 3, 4, 2, <none>, -2
814c814
< srem, 3, 4, 4, <none>, 0
---
> srem, 3, 4, 4, <none>, -4
816c816
< srem, 3, 4, 6, <none>, 0
---
> srem, 3, 4, 6, <none>, -2
819c819
< srem, 3, 5, 1, <none>, 0
---
> srem, 3, 5, 1, <none>, -1
821c821
< srem, 3, 5, 3, <none>, 0
---
> srem, 3, 5, 3, <none>, -3
823c823
< srem, 3, 5, 5, <none>, 0
---
> srem, 3, 5, 5, <none>, -3
825c825
< srem, 3, 5, 7, <none>, 0
---
> srem, 3, 5, 7, <none>, poison
827,828c827,828
< srem, 3, 6, 1, <none>, 0
< srem, 3, 6, 2, <none>, 0
---
> srem, 3, 6, 1, <none>, -1
> srem, 3, 6, 2, <none>, -2
832,833c832,833
< srem, 3, 6, 6, <none>, 0
< srem, 3, 6, 7, <none>, 0
---
> srem, 3, 6, 6, <none>, -2
> srem, 3, 6, 7, <none>, -1
835c835
< srem, 3, 7, 1, <none>, 0
---
> srem, 3, 7, 1, <none>, -1
841c841
< srem, 3, 7, 7, <none>, 0
---
> srem, 3, 7, 7, <none>, -1
```

**Comment by bollu** [comment]
It remains the case that across *all* ops, the diffs are with `sdiv/srem` semantics.

File: SSA/Projects/InstCombine/LLVM/Semantics.lean [hunk 1]
```diff
@@ -68,50 +71,45 @@ def udiv? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
...
 
+The rounding is round to zero, not round to -infty.
+at width 1, -1 / -1 is considered -1
+at width 2, -4 / -1 is considered overflow!
 
--- Probably not a Mathlib worthy name,
--- not sure how you'd mathlibify the precondition
+-/
+def sdiv? {w : Nat} (x y : BitVec w) : Option <| BitVec w :=
+  if y == 0
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/LLVM/Semantics.lean:95]:
```suggestion
  if y == 0 || (w != 1 && x == (intMin w) && y == -1)
```
Let's merge both `if` statements

**Review by alexkeizer — COMMENTED** [review]
LGTM

**Pull request merged**